ros2/diagnostics/src/main.cpp

```diff
@@ -0,0 +1,140 @@
+    Node("diagnostics_publisher"), 
+    timer_count_(0), 
+    last_msg_time_(now())
+    {
+        // Publisher for diagnostic array messages
+        diagnostics_pub_ = create_publisher<diagnostic_msgs::msg::DiagnosticArray>("/diagnostics", 10);
+
+        // Subscriber to the Int16 topic
+        topic_sub_ = create_subscription<std_msgs::msg::Int16>(
+            "/topic", 10, std::bind(&DiagnosticsPublisher::topic_callback, this, _1));
+
+        // Timer to update status based on internal counter
+        timer_ = create_wall_timer(100ms, std::bind(&DiagnosticsPublisher::timer_callback, this));
+
+        // Timer to check for staleness
+        staleness_timer_ = create_wall_timer(500ms, std::bind(&DiagnosticsPublisher::staleness_check_callback, this));
+    }
+
+private:
+    // Callback to get the value and time for the topic
+    void topic_callback(const std_msgs::msg::Int16::SharedPtr msg){
+        value_ = msg->data;
+        last_msg_time_ = now();
+        update_topic_status();
+    }
+
+    // Timer to check if the topic is STALE
+    void staleness_check_callback(){
+        auto now = get_clock()->now();
+        if ((now - last_msg_time_).seconds() > 1.0)
+        {
+            topic_status_.level = diagnostic_msgs::msg::DiagnosticStatus::STALE;
+            topic_status_.message = "No data received for more than 1 second";
+        }
+    }
+
+    // Callback to increase the timer count and publish the diagnostic
+    void timer_callback(){
+        timer_count_ += 0.1;
+
+        update_self_status();
+
+        if (timer_count_ >= 10.0){
+            timer_count_ = 0.0;
+        }
+
+        // Publish diagnostic array
+        diagnostic_msgs::msg::DiagnosticArray diag_array_msg;
+        diag_array_msg.header.stamp = now();
+        diag_array_msg.status.push_back(topic_status_);
+        diag_array_msg.status.push_back(self_status);
+        diagnostics_pub_->publish(diag_array_msg);
+    }
+    
+    // Function to update the Status for the topic 
+    void update_topic_status(){
+        topic_status_.name = "Int16 Topic Status";
+        topic_status_.hardware_id = "int16_topic_001";
+        topic_status_.values.clear();
+        diagnostic_msgs::msg::KeyValue key_value;
+        key_value.key = "Last value";
+        key_value.value = std::to_string(value_);
+        topic_status_.values.push_back(key_value);
+
+        if (value_ <= 1000)
+        {
+            topic_status_.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+            topic_status_.message = "Value within normal range";
+        }
+        else if (value_ <= 1500)
+        {
+            topic_status_.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
+            topic_status_.message = "Value approaching upper limit";
+        }
+        else
+        {
+            topic_status_.level = diagnostic_msgs::msg::DiagnosticStatus::ERROR;
+            topic_status_.message = "Value exceeds upper limit";
+        }
+    }
```

Comment:
The `diagnostics_updater` package seems to be able to do this:

> Send an error if a value is out bounds (e.g. temperature)

Reply:
True, but this tutorial was meant to teach the basics of diagnostics from the ground. I think it's important to understand how the message is build by working with it before using another package that runs everything for you.
Nevertheless, like said, we could work on a second part where we focus on the `diagnostic_updater`  package